shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs

```diff
@@ -67,16 +78,19 @@ genDCerts keys pparams dpState slotWithTTL = do
 -- | Occasionally generate a valid certificate
 genDCert
   :: KeyPairs
-  -> DState
+  -> VrfKeyPairs
+  -> DPState
   -> Gen (Maybe (DCert, KeyPair))
-genDCert keys dState =
-  -- TODO @uroboros Generate _RegPool_ Certificates
+genDCert keys vrfKeys dpState =
   -- TODO @uroboros Generate _RetirePool_ Certificates
   -- TODO @uroboros Generate _Delegate_ Certificates
-  Gen.frequency [ (1, genRegKeyCert keys dState)
-                , (1, genDeRegKeyCert keys dState)
+  Gen.frequency [ (3, genRegKeyCert keys dState)
+                , (4, genDeRegKeyCert keys dState)
```

Comment:
I think we should keep DeReg's at most as likely as RegKey's i.e. `(3, genDeRegKeyCert keys dState)`

Reply:
My rationale had been that since `genDeRegKeyCert` is less likely to have its preconditions met, generating it more often would cause it to have more equivalent %s in coverage. But dropping the weighting to 3 still passes coverage so I’ve done that.

Reply:
@mhuesch @dnadales the return type `Maybe (cert, cert witness key)` was meant to indicate that we may not be able to generate some certs e.g. (as you point out) `DeRegKey` which is only possible if the key was already registered (and if not possible, will not become possible by retrying that particular cert). Returning `mzero` is essentially doing `Gen.discard`, which I was trying to avoid. 

@dnadales I can see this making it harder to reason about the total number of certs in a transaction/trace, but not about particular cert types (e.g. we are indeed asserting that there are at least n DeRegKey certs per trace) . I realise now that I made the assumption that it was more interesting to classify particular types of certs than the total number of them - what kind of reasoning/classification do you think we should do for certs that this approach will prevent?

@mhuesch the code you link to would become something like the following when we generate multiple certs per transaction ... `witnessedCerts <- catMaybes <$> replicateM n genDCert....`

I'm not married to the Maybe return type - just wanting to understand the issue

Reply:
> Returning mzero is essentially doing Gen.discard, which I was trying to avoid.

Why's that?

>  I realise now that I made the assumption that it was more interesting to classify particular types of certs than the total number of them

I don't get this. What are we classifying now? The total number of certs of any type?

I'm fine either way, I just wanted to point out the issue.

Reply:
> I don't get this. What are we classifying now? The total number of certs of any type?

Yes, that's what we've been classifying.

It seems to me like the only benefit of a `Maybe` return type is that it gives us a variable length of list of `DCert`s. Since we'd generate a few `Nothing`s and then they would get filtered out of the final `Seq DCert` return value.

If we use `mzero` / `Gen.discard`, we will always get the same # of `DCert`s I believe, because we'll retry until we get a `Just` value.

It seems like if we want a variable # of `DCert`s we could accomplish that by introduce variance to the length parameter, using a `Hedgehog.Range`. Then we would sort of de-conflate the generation of values from the varying of the length of the list.

Reply:
Today I rewrote these generators to use `Gen.discard` where the `Nothing` values once were. The problem with this approach seems to be that if we are in a scenario where we hit the `discard` branch, we will then hit `discard`s on all subsequent generator attempts, because, for example, the list of `availableKeys` is empty, and will only become non-empty by processing a subsequent signal (thus it will never be possible while we are in this particular evaluation of the `Gen DCert`-typed generator.

So we `discard` until we hit hedgehog's limit and our generator is killed, failing the property.

So it now seems like the `Maybe` approach is an ok alternative. I wonder, though, if we need to [intentionally generate `Nothing` values](https://github.com/input-output-hk/cardano-ledger-specs/blob/cf339d59d09e24005a0b67babce013ed86330e30/shelley/chain-and-ledger/executable-spec/test/Generator/Delegation.hs#L90), which will only be treated as equivalent to "invalid precondition" values, and be dropped.

Reply:
Yes, Gen.discard is no use in the case where valid generation has become impossible (e.g. generating RegKey certs and not having any more keys available to register). Gen.discard will just waste cycles and take longer to fail the property.

We've not been classifying the total number of certs, but rather the proportion (e.g. 1 in 5 transactions in the Ledger trace have a RegKey cert). 

The point of the intentionally generated nothing values you mention above was to serve as "placeholders" - as we add another the next Cert generator to genDCerts (e.g. Delegate certs), we don't want all the existing proportions to shift (e.g. RegKey certs for every 5 transactions) every time we add another cert to the mix. Ultimately (when we're generating all cert types) there should be _no_ Nothings generated intentionally.